### ISSUE TITLE
@sendTo does not work with errorHandler

### ISSUE THREAD

**Issue by priyanshus1** [spring-projects/spring-amqp#1021]
**Affects Version(s):** 2.1.5

## Question
I am using @sendTo annotation with RabbitListenerErrorHandler errorHandler().
But I am getting following error:

```
Caused by: org.springframework.amqp.rabbit.listener.adapter.ReplyFailureException: Failed to send reply with payload 'InvocationResult [returnValue=org.springframework.remoting.support.RemoteInvocationResult@2c663a5c, returnType=null]'
	at org.springframework.amqp.rabbit.listener.adapter.AbstractAdaptableMessageListener.doHandleResult(AbstractAdaptableMessageListener.java:385) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.amqp.rabbit.listener.adapter.AbstractAdaptableMessageListener.handleResult(AbstractAdaptableMessageListener.java:339) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.returnOrThrow(MessagingMessageListenerAdapter.java:162) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.amqp.rabbit.listener.adapter.MessagingMessageListenerAdapter.onMessage(MessagingMessageListenerAdapter.java:147) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.amqp.rabbit.listener.AbstractMessageListenerContainer.doInvokeListener(AbstractMessageListenerContainer.java:1552) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	... 10 common frames omitted
Caused by: org.springframework.amqp.AmqpException: Cannot determine ReplyTo message property value: Request message does not contain reply-to property, and no default response Exchange was set.
	at org.springframework.amqp.rabbit.listener.adapter.AbstractAdaptableMessageListener.getReplyToAddress(AbstractAdaptableMessageListener.java:465) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	at org.springframework.amqp.rabbit.listener.adapter.AbstractAdaptableMessageListener.doHandleResult(AbstractAdaptableMessageListener.java:381) ~[spring-rabbit-2.1.5.RELEASE.jar:2.1.5.RELEASE]
	... 14 common frames omitted
```
## Bug report

When debugged in detail I found out that MessagingMessageListenerAdapter.onMessage() function always create the InvocationResult with null sendTo and null Type; while calling the handleResult().

```java
@Override
	public void onMessage(org.springframework.amqp.core.Message amqpMessage, Channel channel) throws Exception { // NOSONAR
		Message<?> message = toMessagingMessage(amqpMessage);
		if (logger.isDebugEnabled()) {
			logger.debug("Processing [" + message + "]");
		}
		InvocationResult result = null;
		try {
			result = invokeHandler(amqpMessage, channel, message);
			if (result.getReturnValue() != null) {
				handleResult(result, amqpMessage, channel, message);
			}
			else {
				logger.trace("No result object given - no result to handle");
			}
		}
		catch (ListenerExecutionFailedException e) {
			if (this.errorHandler != null) {
				try {
					Object errorResult = this.errorHandler.handleError(amqpMessage, message, e);
					if (errorResult != null) {
						handleResult(new InvocationResult(errorResult, null, null), amqpMessage, channel, message);
					}
					else {
						logger.trace("Error handler returned no result");
					}
				}
				catch (Exception ex) {
					returnOrThrow(amqpMessage, channel, message, ex, ex);
				}
			}
			else {
				returnOrThrow(amqpMessage, channel, message, e.getCause(), e);
			}
		}
	}
```
Well according to documentation mentioned here https://docs.spring.io/spring-amqp/docs/2.1.5.RELEASE/api/ in Interface RabbitListenerErrorHandler
"Handle the error. If an exception is not thrown, the return value is returned to the sender using normal replyTo/@SendTo semantics."

I think we can fix the bug by createing InvocationResult with different Expression value or something.

**Comment by garyrussell** [comment]
Yes, it's a bug; it works if the incoming message has a `replyTo` header, but the send to expression is lost.

**Comment by priyanshus1** [comment]
Hi Gary,

Is it not possible to fix it 2.1.6 ? is it already prepared to ship.
I could also fix it if you want ?

**Comment by artembilan** [comment]
Yes, Gary has marked it to be fixed in the upcoming `2.1.7` version.

And yes: a contribution is more than welcome!

Please, consider to raise PR against `master` and we will back-port it into the `2.1.x` branch.

Thank you!

**Comment by garyrussell** [comment]
It turns out this is only a problem on class-level `@RabbitListener`s.

I went ahead and fixed it because we need to release tomorrow.
